// rn-material-ui-textfield/lib/commonjs/components/outline/styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.borderRadius = void 0;

var _reactNative = require("react-native");

const borderRadius = 4;
exports.borderRadius = borderRadius;
let containerStyle = {
  position: 'absolute',
  overflow: 'hidden'
};

var _default = _reactNative.StyleSheet.create({
  borderLeft: { ..._reactNative.StyleSheet.absoluteFillObject,
    borderRadius,
    right: -borderRadius,
    bottom: -borderRadius
  },
  borderRight: { ..._reactNative.StyleSheet.absoluteFillObject,
    borderRadius,
    left: -borderRadius,
    bottom: -borderRadius
  },
  borderBottom: { ..._reactNative.StyleSheet.absoluteFillObject,
    borderRadius,
    top: -borderRadius,

    /* XXX: Android positioning error workaround */
    bottom: _reactNative.StyleSheet.hairlineWidth
  },
  borderTop: { ..._reactNative.StyleSheet.absoluteFillObject,
    borderRadius,
    left: -borderRadius,
    right: -borderRadius,
    bottom: -borderRadius
  },
  leftContainer: { ...containerStyle,
    top: 0,
    left: 0,
    width: borderRadius,
    height: borderRadius
  },
  rightContainer: { ...containerStyle,
    top: 0,
    right: 0,
    height: borderRadius
  },
  bottomContainer: { ...containerStyle,
    top: borderRadius - (_reactNative.Platform.OS === 'android' ? 0.25 : 0),
    left: 0,
    right: 0,
    bottom: 0
  },
  topContainer: { ...containerStyle,
    top: 0,
    height: borderRadius
  },
  topLineContainer: { ..._reactNative.StyleSheet.absoluteFillObject,
    overflow: 'hidden'
  }
});

exports.default = _default;
